query-frontend: format query start/end/step, log query length

dimitarvdimitrov opened the pull request:
This PR adds another layer of stats to the query-frontend. These stats contain details about the parsed query and its execution. For now these details are only

* start time in milliseconds
* end time in milliseconds
* step in milliseconds

These are logged in query stats with precedence over the raw values received over the API. This makes machine-parsing the values easier.

The PR also logs the time since the param_start and param_end parameters. This will help determine whether the query would hit ingesters or store-gateways. I wasn't sure whether these should be easier to be ready by humans or by machines. I opted for having them as human-readable go duration strings. Loki can also parse these, but I am not sure about other logging solutions.

I plan to open a follow-up PR to also log insights about the portion of the query that was served from the cache.

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

dimitarvdimitrov commented:
here is an example log line

```
ts=2023-10-24T11:24:20.049522817Z caller=handler.go:317 level=info user=16517 msg="query stats" component=query-frontend method=GET path=/prometheus/api/v1/query_range user_agent=Grafana/10.1.0 response_time=4.508066ms response_size_bytes=189 query_wall_time_seconds=0.000148441 fetched_series_count=0 fetched_chunk_bytes=0 fetched_chunks_count=0 fetched_index_bytes=0 sharded_queries=0 split_queries=1 estimated_series_count=0 param_end=1698146640000 param_query=1 param_start=1698146340000 param_step=60000 length=5m0s time_since_param_start=5m20.049507292s time_since_param_end=20.049507829s cached_results_bytes=0 uncached_results_bytes=216 header_x_cloud_org_id=916 header_x_access_policy_id=5a184398-a9d4-4c6f-8403-bb9188a9fdb7 header_fromalert=true header_x_forwarded_for="34.69.37.85, 34.110.187.6, 130.211.1.69" status=success
```
and trimmed
```
msg="query stats" param_end=1698146640000 param_query=1 param_start=1698146340000 param_step=60000 length=5m0s time_since_param_start=5m20.049507292s time_since_param_end=20.049507829s
```

charleskorn commented on pkg/frontend/transport/handler.go:
```diff
@@ -302,9 +315,25 @@ func (f *Handler) reportQueryStats(r *http.Request, queryString url.Values, quer
+			var millis int64
+			switch k {
+			case "start":
+				millis = details.Start.UnixMilli()
+			case "end":
+				millis = details.End.UnixMilli()
```
What if we emitted these in RFC 3339 format? Then they'd be human and machine readable.

charleskorn commented on pkg/frontend/transport/handler.go:
```diff
@@ -302,9 +315,25 @@ func (f *Handler) reportQueryStats(r *http.Request, queryString url.Values, quer
 	for k, v := range queryString {
-		fields = append(fields, fmt.Sprintf("param_%s", k), strings.Join(v, ","))
+		var formattedValue string
+		if details != nil && (k == "start" || k == "end" || k == "step") {
```
Should we apply the same formatting to the `time` param for instant queries, for consistency?

charleskorn commented on pkg/frontend/transport/handler_test.go:
```diff
@@ -178,7 +178,7 @@ func TestHandler_ServeHTTP(t *testing.T) {
 				msg := logger.logMessages[0]
-				require.Len(t, msg, 18+len(tt.expectedParams))
+				require.Len(t, msg, 21+len(tt.expectedParams))
 				require.Equal(t, level.InfoValue(), msg["level"])
```
Should we add a test case here that exercises the formatting for `start`, `end` and `step`?

dimitarvdimitrov replied:
i added that in https://github.com/grafana/mimir/pull/6473/commits/8e6a8f4c6090fb2bea6b9088aa53d3b864cbb0ad

pracucci commented on pkg/frontend/transport/handler.go:
```diff
@@ -277,7 +282,15 @@ func (f *Handler) reportQueryStats(r *http.Request, queryString url.Values, quer
+
+	if details != nil {
+		logMessage = append(logMessage,
+			"length", details.End.Sub(details.Start).String(),
+			"time_since_param_start", time.Since(details.Start).String(),
+			"time_since_param_end", time.Since(details.End).String(),
```
This block is code assumes there's always start/end, but we have query APIs without that (e.g. /metadata). It could be misleading logging these values for API endpoints where start/end has no meaning.

dimitarvdimitrov replied:
I agree about the /metadata requests. I removed the log lines in case the times are unset.

But I'm not sure whether having different log fields for `time_since*` will be helpful. The point of these is to ease spotting changes in query patterns. The loki queries will become more complicated if we have to account for multiple fields. I think changing them to instead `time_to_min_time` and `time_to_max_time` will make more sense.

This requires the changes you suggested in the other comment for min/max time. In the meantime I can remove the `time_since_param_end` and `time_since_param_start` log fields. 


WDYT?

dimitarvdimitrov replied:
i added `time_since_max_time` and `time_since_min_time` and `length` all based on the min/max times of the query instead of start/end.

pracucci commented on pkg/frontend/querymiddleware/stats.go:
```diff
@@ -62,5 +65,41 @@ func (s queryStatsMiddleware) Do(ctx context.Context, req Request) (Response, er
 	}
 
+	if details := QueryDetailsFromContext(ctx); details != nil {
+		details.Start = time.UnixMilli(req.GetStart())
+		details.End = time.UnixMilli(req.GetEnd())
+		details.Step = time.Duration(req.GetStep()) * time.Millisecond
```
I think there's a fondamental flaw here. According to the PR description, start/end are read to log the actual query time range and figure out whether a query only hits ingesters or store-gateway. However, the actual query time range is not just driven by start/end parameter but also by vector selector duration (e.g. think about an instant query where start=end=now but has `metric[30d]`).

I think a better approach may be storing the **min and max timestamp** in the querier stats, and then use such information to log the actual query timerange (which is a very interesting data point we're completely missing today).

dimitarvdimitrov replied:
> I think a better approach may be storing the min and max timestamp in the querier stats, and then use such information to log the actual query timerange (which is a very interesting data point we're completely missing today).

i agree. Looking at the code which currently validates this I think we have to expose a method of the promQL engine https://github.com/grafana/mimir/blob/423e6f00a9f116e2fe5e0e318ac4fabd7cc5dcf3/vendor/github.com/prometheus/prometheus/promql/engine.go#L820

I will propose this upstream first.

In the meantime is it ok to merge the current approximation? It will be a few minutes off, but it will have the correct ballpark value

dimitarvdimitrov replied:
I opened https://github.com/prometheus/prometheus/pull/13085 to export `findMinMaxTime`

dimitarvdimitrov replied:
that got merged and into mimir quicker than I expected. I'm not integrating it in this PR

dimitarvdimitrov commented:
@pracucci can you take another look at this PR when you can?

dimitarvdimitrov commented:
@charleskorn can you take a look at this RP please? I think I've addressed all comments you two had. This will also unblock merging https://github.com/grafana/mimir/pull/6477
